Fix #1948

hroff-1902 opened the pull request:
How to reproduce: remove the item with key="0" from the minimal_roi dict.

* resolves #1948 
* test added
* code slightly improved (return value in min_roi_reached)
* wrong comment fixed

xmatthias commented on freqtrade/strategy/interface.py:
```diff
@@ -357,13 +357,13 @@ def min_roi_reached(self, trade: Trade, current_profit: float, current_time: dat
+        roi_list = list(filter(lambda x: x <= trade_dur, self.minimal_roi.keys()))
+        if not roi_list:
+            return False
+        roi_entry = max(roi_list)
```
This whole section should be extracted into it's own function and called from both here and `backtesting`.

hroff-1902 replied:
Maybe a better option would be to return the roi_entry just calculated in the should_sell() function a few lines above... (in a new member of SellCheckTuple, like SellCheckTuple.data; for SellType.ROI it can be assigned roi_entry from min_roi_reached(), for SellType.STOP_LOSS and SellType.TRAILING_STOP_LOSS -- the trade.stop_loss as used in stop_loss_reached() or some other useful value...

What do you think?

Also, the comment here
```
# get next entry in min_roi > to trade duration
# Interface.py skips on trade_duration <= duration
```
says the value of min_roi here should be different from one in strategy/interface.py.
And it also contradicts with the code implemented. Maybe someone has fixed an error here but did not change the old comment. Could you review these both places where roi_entry is calculated?

xmatthias replied:
> # Interface.py skips on trade_duration <= duration

IIRC, the 2nd comment was there since previously this was done differently in interface.py - but we discovered that this can lead to odd behaviour if a unsorted ROI dict is passed in, so the 2 methods were aligned.
Probably the comment was forgotten

hroff-1902 replied:
Done (duplicated code extracted into a separate function)

hroff-1902 commented:
Or, as an option, we could check the consistency of the min_roi at startup and reject it if it does not contain an entry with the '0' key... (another option is to add something for this zero-minutes entry)

xmatthias commented on freqtrade/optimize/backtesting.py:
```diff
@@ -253,22 +253,19 @@ def _get_sell_trade_entry(
+                        closerate = - (trade.open_rate * roi + trade.open_rate *
+                                       (1 + trade.fee_open)) / (trade.fee_close - 1)
+                    else:
+                        closerate = sell_row.open
```
I don't think this line (or the else branch in total) can be hit at all.
Remember that it's within the `if sell.sell_type == (SellType.ROI):` block.

Hitting this line would assume that `should_sell()` reports sell-reason as ROI - but `min_roi_reached_entry()` does not have a ROI for this "open" time - so this would indicate a huge bug somewhere in should_sell (which i don't think is there atm), since `should_sell()` uses `min_roi_reached_entry()` too.

I just tested this by raising an exception on this line (and trying different ROI configurations, including an empty dict) - and this was never hit, so I'm not sure if we shouldn't remove this line completely - since it's unreachable code.

hroff-1902 replied:
> I don't think this line (or the else branch in total) can be hit at all.

It will be hit in exactly same conditions when the user got previously exception in #1948 
(previously it was an unexpected exception generated by max(), now it's None returned as min_roi entry val, it's handled above in the calling function, so some closerate should be assigned to proceed...)

xmatthias replied:
maybe i misunderstand #1948 - but wasn't that a ROI dict without a 0 entry:
e.g.

```python
minimal_roi = { "200": "0.01"}
```

hroff-1902 replied:
> maybe i misunderstand #1948 - but wasn't that a ROI dict without a 0 entry:
> e.g.
> 
> ```python
> minimal_roi = { "200": "0.01"}
> ```

Yes and that line assigning the `closerate` should be hit when `trade_dur < 200`, i.e. when `None` is returned from `self.strategy.min_roi_reached_entry(trade_dur)`. I assume.

xmatthias replied:
well that's true for the `interface.py` file - but not here, since this whole code-section is ONLY hit if the sell-reason is ROI.

So if tradedur < 200 - then there is 
a) no sell or
b)  sell-reason is something else (stoploss / sell-signal, ...)

however this part won't be reached since for that to happen, you'd have to have sell_reason=ROI + trade_dur < 200 - (but then sell-reason won't be ROI)...

hroff-1902 commented:
Strange, I cannot get the difference in behaviour... When I run the current code from develop, it raises exception for both main freqtrade and for backtesting:
```
(.env) $ git status
On branch minor-typos-1
Your branch is up to date with 'origin/minor-typos-1'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	config.json.1948

nothing added to commit but untracked files present (use "git add" to track)
(.env) $ git checkout develop 
Switched to branch 'develop'
Your branch is up to date with 'origin/develop'.
(.env) $ diff config_binance.json.example config.json.example.1948 
12a13,17
>     "minimal_roi": {
>         "40":  0.0,
>         "30":  0.01,
>         "20":  0.02
>     },
(.env) $ ./freqtrade/main.py -c config.json.example.1948
2019-06-26 22:35:10,339 - freqtrade.configuration - INFO - Verbosity set to 0
2019-06-26 22:35:10,339 - freqtrade.configuration - INFO - Dry run is enabled
2019-06-26 22:35:10,339 - freqtrade.configuration - INFO - Using DB: "sqlite://"
2019-06-26 22:35:10,339 - freqtrade.configuration - INFO - Checking exchange...
2019-06-26 22:35:10,339 - freqtrade.configuration - INFO - Exchange "binance" is officially supported by the Freqtrade development team.
2019-06-26 22:35:10,340 - freqtrade.configuration - INFO - Using max_open_trades: 3 ...
2019-06-26 22:35:10,340 - freqtrade.configuration - INFO - Using data folder: user_data/data/binance ...
2019-06-26 22:35:10,340 - freqtrade.freqtradebot - INFO - Starting freqtrade 2019.6-dev
2019-06-26 22:35:10,347 - freqtrade.resolvers.strategy_resolver - INFO - Using resolved strategy DefaultStrategy from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy'
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'minimal_roi' with value in config file: {'40': 0.0, '30': 0.01, '20': 0.02}.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ticker_interval' with value in config file: 5m.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'trailing_stop' with value in config file: False.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: BTC.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: 0.05.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'use_sell_signal' with value in config file: False.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'sell_profit_only' with value in config file: False.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ignore_roi_if_buy_signal' with value in config file: False.
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'40': 0.0, '30': 0.01, '20': 0.02}
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ticker_interval: 5m
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.1
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: False
2019-06-26 22:35:10,348 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.0
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: False
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: False
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'buy': 'limit', 'sell': 'limit', 'stoploss': 'limit', 'stoploss_on_exchange': False}
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'buy': 'gtc', 'sell': 'gtc'}
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: BTC
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: 0.05
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_sell_signal: False
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using sell_profit_only: False
2019-06-26 22:35:10,349 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_buy_signal: False
2019-06-26 22:35:10,349 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2019-06-26 22:35:10,349 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-06-26 22:35:10,351 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 200}
2019-06-26 22:35:10,362 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2019-06-26 22:35:11,940 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange Binance
2019-06-26 22:35:11,941 - freqtrade.wallets - INFO - Wallets synced.
2019-06-26 22:35:11,941 - freqtrade.resolvers.pairlist_resolver - WARNING - Path "user_data/pairlist" does not exist
2019-06-26 22:35:11,945 - freqtrade.resolvers.pairlist_resolver - INFO - Using resolved pairlist StaticPairList from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/pairlist'
2019-06-26 22:35:11,957 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'running'}
2019-06-26 22:35:11,957 - freqtrade.worker - INFO - Changing state to: RUNNING
2019-06-26 22:35:11,958 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': warning, 'status': 'Dry run is enabled. All trades are simulated.'}
2019-06-26 22:35:11,958 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': custom, 'status': "*Exchange:* `binance`\n*Stake per trade:* `0.05 BTC`\n*Minimum ROI:* `{'40': 0.0, '30': 0.01, '20': 0.02}`\n*Stoploss:* `-0.1`\n*Ticker Interval:* `5m`\n*Strategy:* `DefaultStrategy`"}
2019-06-26 22:35:11,958 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': "Searching for BTC pairs to buy and sell based on StaticPairList: ['AST/BTC', 'ETC/BTC', 'ETH/BTC', 'EOS/BTC', 'IOTA/BTC', 'LTC/BTC', 'MTH/BTC', 'NCASH/BTC', 'TNT/BTC', 'XMR/BTC', 'XLM/BTC', 'XRP/BTC']"}
2019-06-26 22:35:15,429 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 0.05 ...
2019-06-26 22:35:15,429 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2019-06-26 22:35:15,720 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'AST/BTC', 'limit': 3.98e-06, 'order_type': 'limit', 'stake_amount': 0.05, 'stake_currency': 'BTC', 'fiat_currency': 'USD'}
2019-06-26 22:35:15,724 - freqtrade.wallets - INFO - Wallets synced.
2019-06-26 22:35:15,734 - freqtrade.wallets - INFO - Wallets synced.
2019-06-26 22:35:16,968 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=1, pair=AST/BTC, amount=12562.81407035, open_rate=0.00000398, open_since=just now)
2019-06-26 22:35:16,970 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=1, pair=AST/BTC, amount=12562.81407035, open_rate=0.00000398, open_since=just now) failed: myTrade-Dict empty found
2019-06-26 22:35:16,970 - freqtrade.persistence - INFO - Updating trade (id=1) ...
2019-06-26 22:35:16,970 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=1, pair=AST/BTC, amount=12562.81407035, open_rate=0.00000398, open_since=just now).
2019-06-26 22:35:17,267 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "./freqtrade/main.py", line 52, in main
    worker.run()
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/worker.py", line 72, in run
    state = self._worker(old_state=state)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/worker.py", line 110, in _worker
    self._throttle(func=self._process, min_secs=throttle_secs)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/worker.py", line 123, in _throttle
    result = func(*args, **kwargs)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/worker.py", line 133, in _process
    state_changed = self.freqtrade.process()
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/freqtradebot.py", line 135, in process
    state_changed |= self.process_maybe_execute_sell(trade)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/freqtradebot.py", line 458, in process_maybe_execute_sell
    result = self.handle_trade(trade)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/freqtradebot.py", line 595, in handle_trade
    if self.check_sell(trade, sell_rate, buy, sell):
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/freqtradebot.py", line 716, in check_sell
    should_sell = self.strategy.should_sell(trade, sell_rate, datetime.utcnow(), buy, sell)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy/interface.py", line 280, in should_sell
    if self.min_roi_reached(trade=trade, current_profit=current_profit, current_time=date):
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy/interface.py", line 361, in min_roi_reached
    roi_entry = max(list(filter(lambda x: trade_dur >= x, self.minimal_roi.keys())))
ValueError: max() arg is an empty sequence
2019-06-26 22:35:17,268 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2019-06-26 22:35:17,268 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2019-06-26 22:35:17,268 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
(.env) $ ./freqtrade/main.py -c config.json.example.1948 backtesting
2019-06-26 22:35:29,543 - freqtrade.configuration - INFO - Verbosity set to 0
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Dry run is enabled
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Using DB: "sqlite://"
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Checking exchange...
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Exchange "binance" is officially supported by the Freqtrade development team.
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Using max_open_trades: 3 ...
2019-06-26 22:35:29,544 - freqtrade.configuration - INFO - Using data folder: user_data/data/binance ...
2019-06-26 22:35:29,545 - freqtrade.configuration - INFO - Storing backtest results to user_data/backtest_data/backtest-result.json ...
2019-06-26 22:35:29,545 - freqtrade.optimize - INFO - Starting freqtrade in Backtesting mode
2019-06-26 22:35:29,545 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2019-06-26 22:35:29,545 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-06-26 22:35:29,547 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 200}
2019-06-26 22:35:29,558 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2019-06-26 22:35:30,503 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange Binance
2019-06-26 22:35:30,507 - freqtrade.resolvers.strategy_resolver - INFO - Using resolved strategy DefaultStrategy from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy'
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'minimal_roi' with value in config file: {'40': 0.0, '30': 0.01, '20': 0.02}.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ticker_interval' with value in config file: 5m.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'trailing_stop' with value in config file: False.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: BTC.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: 0.05.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'use_sell_signal' with value in config file: False.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'sell_profit_only' with value in config file: False.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ignore_roi_if_buy_signal' with value in config file: False.
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'40': 0.0, '30': 0.01, '20': 0.02}
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ticker_interval: 5m
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.1
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: False
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.0
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: False
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: False
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'buy': 'limit', 'sell': 'limit', 'stoploss': 'limit', 'stoploss_on_exchange': False}
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'buy': 'gtc', 'sell': 'gtc'}
2019-06-26 22:35:30,508 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: BTC
2019-06-26 22:35:30,509 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: 0.05
2019-06-26 22:35:30,509 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_sell_signal: False
2019-06-26 22:35:30,509 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using sell_profit_only: False
2019-06-26 22:35:30,509 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_buy_signal: False
2019-06-26 22:35:30,509 - freqtrade.optimize.backtesting - INFO - Using stake_currency: BTC ...
2019-06-26 22:35:30,509 - freqtrade.optimize.backtesting - INFO - Using stake_amount: 0.05 ...
2019-06-26 22:35:30,509 - freqtrade.data.history - INFO - Using local backtesting data ...
2019-06-26 22:35:30,600 - freqtrade.data.converter - INFO - Missing data fillup for AST/BTC: before: 8618 - after: 8738
2019-06-26 22:35:30,660 - freqtrade.data.converter - INFO - Missing data fillup for ETC/BTC: before: 13115 - after: 13247
2019-06-26 22:35:30,719 - freqtrade.data.converter - INFO - Missing data fillup for ETH/BTC: before: 13115 - after: 13247
2019-06-26 22:35:30,770 - freqtrade.data.converter - INFO - Missing data fillup for EOS/BTC: before: 8613 - after: 8733
2019-06-26 22:35:30,815 - freqtrade.data.converter - INFO - Missing data fillup for IOTA/BTC: before: 8611 - after: 8731
2019-06-26 22:35:30,874 - freqtrade.data.converter - INFO - Missing data fillup for LTC/BTC: before: 13115 - after: 13247
2019-06-26 22:35:30,921 - freqtrade.data.converter - INFO - Missing data fillup for MTH/BTC: before: 8608 - after: 8728
2019-06-26 22:35:30,966 - freqtrade.data.converter - INFO - Missing data fillup for NCASH/BTC: before: 8607 - after: 8727
2019-06-26 22:35:31,017 - freqtrade.data.converter - INFO - Missing data fillup for TNT/BTC: before: 8599 - after: 8719
2019-06-26 22:35:31,108 - freqtrade.data.converter - INFO - Missing data fillup for XMR/BTC: before: 13115 - after: 13247
2019-06-26 22:35:31,166 - freqtrade.data.converter - INFO - Missing data fillup for XLM/BTC: before: 12330 - after: 12462
2019-06-26 22:35:31,214 - freqtrade.data.converter - INFO - Missing data fillup for XRP/BTC: before: 8597 - after: 8717
2019-06-26 22:35:31,221 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2019-04-27T23:30:00+00:00 up to 2019-06-13T23:50:00+00:00 (47 days)..
2019-06-26 22:35:31,221 - freqtrade.optimize.backtesting - INFO - Running backtesting for Strategy DefaultStrategy
2019-06-26 22:35:35,499 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "./freqtrade/main.py", line 46, in main
    args.func(args)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/optimize/__init__.py", line 57, in start_backtesting
    backtesting.start()
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/optimize/backtesting.py", line 447, in start
    'end_date': max_date,
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/optimize/backtesting.py", line 377, in backtest
    max_open_trades)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/optimize/backtesting.py", line 253, in _get_sell_trade_entry
    sell_row.sell, low=sell_row.low, high=sell_row.high)
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy/interface.py", line 280, in should_sell
    if self.min_roi_reached(trade=trade, current_profit=current_profit, current_time=date):
  File "/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy/interface.py", line 361, in min_roi_reached
    roi_entry = max(list(filter(lambda x: trade_dur >= x, self.minimal_roi.keys())))
ValueError: max() arg is an empty sequence
```

I assume then, the `closerate = sell_row.open` line we discussing above should also be hit in the same conditions, but assert is not called:

```
(.env) $ git checkout fix/1948 
Switched to branch 'fix/1948'
Your branch is up to date with 'origin/fix/1948'.
(.env) $ vi freqtrade/optimize/backtesting.py
(.env) $ git diff
diff --git a/freqtrade/optimize/backtesting.py b/freqtrade/optimize/backtesting.py
index e0660eac..b53ea7f4 100644
--- a/freqtrade/optimize/backtesting.py
+++ b/freqtrade/optimize/backtesting.py
@@ -265,6 +265,7 @@ class Backtesting(object):
                         closerate = - (trade.open_rate * roi + trade.open_rate *
                                        (1 + trade.fee_open)) / (trade.fee_close - 1)
                     else:
+                        assert False, "Should not be reached!"
                         closerate = sell_row.open
                 else:
                     closerate = sell_row.open
(.env) $ ./freqtrade/main.py -c config.json.example.1948 backtesting
2019-06-26 22:42:45,937 - freqtrade.configuration - INFO - Verbosity set to 0
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Dry run is enabled
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Using DB: "sqlite://"
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Checking exchange...
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Exchange "binance" is officially supported by the Freqtrade development team.
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Using max_open_trades: 3 ...
2019-06-26 22:42:45,938 - freqtrade.configuration - INFO - Using data folder: user_data/data/binance ...
2019-06-26 22:42:45,939 - freqtrade.configuration - INFO - Storing backtest results to user_data/backtest_data/backtest-result.json ...
2019-06-26 22:42:45,939 - freqtrade.optimize - INFO - Starting freqtrade in Backtesting mode
2019-06-26 22:42:45,939 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2019-06-26 22:42:45,939 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-06-26 22:42:45,941 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 200}
2019-06-26 22:42:45,953 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2019-06-26 22:42:46,916 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange Binance
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Using resolved strategy DefaultStrategy from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/strategy'
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'minimal_roi' with value in config file: {'40': 0.0, '30': 0.01, '20': 0.02}.
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ticker_interval' with value in config file: 5m.
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'trailing_stop' with value in config file: False.
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: BTC.
2019-06-26 22:42:46,921 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: 0.05.
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'use_sell_signal' with value in config file: False.
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'sell_profit_only' with value in config file: False.
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ignore_roi_if_buy_signal' with value in config file: False.
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'40': 0.0, '30': 0.01, '20': 0.02}
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ticker_interval: 5m
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.1
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: False
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.0
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: False
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: False
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'buy': 'limit', 'sell': 'limit', 'stoploss': 'limit', 'stoploss_on_exchange': False}
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'buy': 'gtc', 'sell': 'gtc'}
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: BTC
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: 0.05
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_sell_signal: False
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using sell_profit_only: False
2019-06-26 22:42:46,922 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_buy_signal: False
2019-06-26 22:42:46,922 - freqtrade.optimize.backtesting - INFO - Using stake_currency: BTC ...
2019-06-26 22:42:46,922 - freqtrade.optimize.backtesting - INFO - Using stake_amount: 0.05 ...
2019-06-26 22:42:46,923 - freqtrade.data.history - INFO - Using local backtesting data ...
2019-06-26 22:42:47,580 - freqtrade.optimize.backtesting - INFO - Running backtesting for Strategy DefaultStrategy
2019-06-26 22:42:47,586 - freqtrade.data.history - WARNING - AST/BTC has missing frames: expected 13540, got 8738, that's 4802 missing values
2019-06-26 22:42:47,586 - freqtrade.data.history - WARNING - ETC/BTC has missing frames: expected 13540, got 13247, that's 293 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - ETH/BTC has missing frames: expected 13540, got 13247, that's 293 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - EOS/BTC has missing frames: expected 13540, got 8733, that's 4807 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - IOTA/BTC has missing frames: expected 13540, got 8731, that's 4809 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - LTC/BTC has missing frames: expected 13540, got 13247, that's 293 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - MTH/BTC has missing frames: expected 13540, got 8728, that's 4812 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - NCASH/BTC has missing frames: expected 13540, got 8727, that's 4813 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - TNT/BTC has missing frames: expected 13540, got 8719, that's 4821 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - XMR/BTC has missing frames: expected 13540, got 13247, that's 293 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - XLM/BTC has missing frames: expected 13540, got 12462, that's 1078 missing values
2019-06-26 22:42:47,587 - freqtrade.data.history - WARNING - XRP/BTC has missing frames: expected 13540, got 8717, that's 4823 missing values
2019-06-26 22:42:47,587 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2019-04-27T23:30:00+00:00 up to 2019-06-13T23:50:00+00:00 (47 days)..
Result for strategy DefaultStrategy
========================================================= BACKTESTING REPORT ========================================================
| pair      |   buy count |   avg profit % |   cum profit % |   tot profit BTC |   tot profit % | avg duration   |   profit |   loss |
|:----------|------------:|---------------:|---------------:|-----------------:|---------------:|:---------------|---------:|-------:|
| AST/BTC   |          33 |          -0.31 |         -10.37 |      -0.00519000 |          -3.46 | 2:00:00        |        7 |      2 |
| ETC/BTC   |          37 |          -0.39 |         -14.37 |      -0.00719000 |          -4.79 | 7:31:00        |        5 |      2 |
| ETH/BTC   |          22 |          -0.44 |          -9.63 |      -0.00482025 |          -3.21 | 23:41:00       |        5 |      2 |
| EOS/BTC   |          14 |           0.00 |           0.00 |       0.00000000 |           0.00 | 2:05:00        |        0 |      0 |
| IOTA/BTC  |          20 |           0.25 |           5.00 |       0.00250000 |           1.67 | 1:56:00        |        3 |      0 |
| LTC/BTC   |          33 |          -0.03 |          -0.91 |      -0.00045354 |          -0.30 | 4:17:00        |        2 |      1 |
| MTH/BTC   |          24 |          -0.90 |         -21.55 |      -0.01078500 |          -7.18 | 4:09:00        |        5 |      3 |
| NCASH/BTC |          20 |          -0.83 |         -16.55 |      -0.00828500 |          -5.52 | 5:09:00        |        8 |      3 |
| TNT/BTC   |          22 |          -0.14 |          -3.19 |      -0.00159500 |          -1.06 | 4:03:00        |        5 |      1 |
| XMR/BTC   |          55 |           0.01 |           0.50 |       0.00025172 |           0.17 | 2:09:00        |        3 |      1 |
| XLM/BTC   |          20 |          -1.67 |         -33.46 |      -0.01674435 |         -11.15 | 17:50:00       |        2 |      4 |
| XRP/BTC   |          13 |          -1.49 |         -19.36 |      -0.00969000 |          -6.45 | 23:46:00       |        1 |      2 |
| TOTAL     |         313 |          -0.40 |        -123.88 |      -0.06200142 |         -41.29 | 6:52:00        |       46 |     21 |
========================================================= SELL REASON STATS =========================================================
| Sell Reason   |   Count |
|:--------------|--------:|
| roi           |     292 |
| stop_loss     |      17 |
| force_sell    |       4 |
====================================================== LEFT OPEN TRADES REPORT ======================================================
| pair    |   buy count |   avg profit % |   cum profit % |   tot profit BTC |   tot profit % | avg duration      |   profit |   loss |
|:--------|------------:|---------------:|---------------:|-----------------:|---------------:|:------------------|---------:|-------:|
| ETH/BTC |           1 |          -5.45 |          -5.45 |      -0.00272525 |          -1.82 | 14 days, 19:05:00 |        0 |      1 |
| LTC/BTC |           1 |          -2.90 |          -2.90 |      -0.00145354 |          -0.97 | 7:05:00           |        0 |      1 |
| XMR/BTC |           1 |          -2.49 |          -2.49 |      -0.00124828 |          -0.83 | 5:50:00           |        0 |      1 |
| XLM/BTC |           1 |          -4.91 |          -4.91 |      -0.00245935 |          -1.64 | 3 days, 13:30:00  |        0 |      1 |
| TOTAL   |           4 |          -3.94 |         -15.76 |      -0.00788642 |          -5.25 | 4 days, 17:22:00  |        0 |      4 |
```

Did not get the reason yet, because this PR is intended to fix #1948 and the code in the `min_roi_reached_entry()` should return None in exactly same conditions when it raised an exception before...

xmatthias commented:
technically, you are correct, it does return None in the same conditions - but the same conditions don't happen.

Let me try to explain this (it's not directly in the change, but in the process how things are called):
During backtesting `min_roi_reached_entry()` is called from 2 places.
- If it's called from `interface.py`, then it'll return `None` (or the value) - depending on the time
- if it's called from `backtesting.py` - then the first call did already happen, returned a value (not `None`!) and `min_roi_reached` returned true (which can only happen if `min_roi_reached_entry()` returned a value). This is the ONLY case where `min_roi_reached_entry()` is ever called from backtesting - in which case `None` can never be a return value.

case 1 (`interface.py`) which is where the exception from the issue occurs, step2 is skipped because the sell-reason is different.

hroff-1902 commented:
> During backtesting `min_roi_reached_entry()` is called from 2 places.

Duh, now I got what I was missing here...

Thanks for you time, for the detailed explaination. 👍

hroff-1902 commented:
Can be merged?
